platform/rpc/nats/bus.go

```diff
@@ -0,0 +1,246 @@
+package nats
+
+import (
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/nats-io/nats"
+
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/platform"
+	"github.com/weaveworks/flux/platform/rpc"
+)
+
+const (
+	timeout      = 5 * time.Second
+	presenceTick = 50 * time.Millisecond
+	encoderType  = nats.JSON_ENCODER
+
+	methodPing         = ".Platform.Ping"
+	methodAllServices  = ".Platform.AllServices"
+	methodSomeServices = ".Platform.SomeServices"
+	methodRegrade      = ".Platform.Regrade"
+)
+
+type NATS struct {
+	url string
+	// It's convenient to send (or request) on an encoding connection,
+	// since that'll do encoding work for us. When receiving though,
+	// we want to decode based on the method as given in the subject,
+	// so we use a regular connection and do the decoding ourselves.
+	snd *nats.EncodedConn
+	rcv *nats.Conn
+}
+
+var _ platform.MessageBus = &NATS{}
+
+func NewMessageBus(url string) (*NATS, error) {
+	conn, err := nats.Connect(url)
+	if err != nil {
+		return nil, err
+	}
+	encConn, err := nats.NewEncodedConn(conn, encoderType)
+	if err != nil {
+		return nil, err
+	}
+	return &NATS{
+		url: url,
+		rcv: conn,
+		snd: encConn,
+	}, nil
+}
+
+// Wait up to `timeout` for a particular instance to connect. Mostly
+// useful for synchronising during testing.
+func (n *NATS) AwaitPresence(instID flux.InstanceID, timeout time.Duration) error {
+	timer := time.After(timeout)
+	attempts := time.NewTicker(presenceTick)
+	defer attempts.Stop()
+
+	for {
+		select {
+		case <-attempts.C:
+			if err := n.Ping(instID); err == nil {
+				return nil
+			}
+		case <-timer:
+			return errors.New("presence timeout")
+		}
+	}
+}
+
+func (n *NATS) Ping(instID flux.InstanceID) error {
+	var response PingResponse
+	err := n.snd.Request(string(instID)+methodPing, ping{}, &response, timeout)
+	if err == nil {
+		err = maybeError(response.Error)
+	}
+	return err
+}
+
+type AllServicesResponse struct {
+	Services []platform.Service
+	Error    string
+}
+
+type SomeServicesResponse struct {
+	Services []platform.Service
+	Error    string
+}
+
+type RegradeResponse struct {
+	Result rpc.RegradeResult
+	Error  string
+}
+
+type ping struct{}
+
+type PingResponse struct {
+	Error string
+}
+
+func maybeError(msg string) error {
+	if msg != "" {
+		return errors.New(msg)
+	}
+	return nil
+}
+
+func maybeString(err error) string {
+	if err != nil {
+		return err.Error()
+	}
+	return ""
+}
```

Comment:
Well, if they were conversion functions, then perhaps; but they can also produce nil errors, which is a change in semantic meaning, so that's also misleading.

I'm telling you that

```go
func foo() {
    // ...
    return stringAsError(response.Error)
}
```

is somewhat confusing and nonidiomatic, and asking you to please write it out as

```go
func foo() {
    // ...
    if response.Error != "" {
        return errors.New(response.Error)
    }
    return nil
}
```

at each site.